rclcpp/include/rclcpp/timer.hpp

```diff
@@ -229,6 +232,11 @@ class GenericTimer : public TimerBase
     TRACEPOINT(callback_start, reinterpret_cast<const void *>(&callback_), false);
     execute_callback_delegate<>();
     TRACEPOINT(callback_end, reinterpret_cast<const void *>(&callback_));
+    if (amount_of_callbacks_ != 0) {
+      if (amount_of_callbacks_ == ++callbacks_called_) {
```

Comment:
`callbacks_called_` will never be reset after complete the `amount_of_callbacks_`?
in this case, user application must destroy this timer object to run the same timer again.
besides,if user cancels the timer along the way, user should be able to know how many times left when it reset(restart)?

Or, if `calcel` is called, it will reset `ccallbacks_called_`? this could be better user experience and consistent behavior?

Reply:
I see your point. At the moment what happens is that `callbacks_called_` is not being modified by the `reset()` method, that would imply that the timer would go to an ongoing state and then cancel itself due to the conditions. I drafted a simple workflow to ilustrate what is happening right now.
![image](https://user-images.githubusercontent.com/51831786/192061761-06df702f-141a-4d0e-aa3e-b8e34ad0392f.png)

I think this actually doesn't make sense, thinking from the user point of view, what I would expect if I reset a "one-shot timer" for example is to get the same "one-shot timer" that triggers once. I think the workflow of the feature should be something like the following:
![image](https://user-images.githubusercontent.com/51831786/192062168-4b0191da-8d7e-4e10-8306-80e779063cde.png)

Taking that into account I would propose two possible solutions to get that behavior. One would be to reset the variable `callbacks_called_` after canceling the timer:
```
if (amount_of_callbacks_ != 0) {
    if (amount_of_callbacks_ <= ++callbacks_called_) {
      cancel();
      callbacks_called = 0;
    }
}
```
An the other option I would propose is to reset the value of `callbacks_called` on the `reset()` method, that would mean to modify that method to something like this:
```
void
TimerBase::reset()
{
  rcl_ret_t ret = rcl_timer_reset(timer_handle_.get());
  callbacks_called = 0;
  if (ret != RCL_RET_OK) {
    rclcpp::exceptions::throw_from_rcl_error(ret, "Couldn't reset timer");
  }
}
```

Let me know what you think regarding the behavior this feature should have after using `reset()` on the timer and how we could accomplish that.

Reply:
@Voldivh really appreciate the detailed flow diagram and explanation, this takes time but really good for understanding for everyone 🎉 

> reset the value of callbacks_called on the reset() method

i would take this proposal,

- `reset` makes more sense in general semantic? it actually `reset` the timer. (i may be wrong, I am not native English speaker... 😅  )
- probably it would be nice to keep the `state` after `cancel` for more extension for future, such as interface to let user application to access the `state` of the timer. (like how many times has callback been fired on this timer?)

Reply:
I had to modify add a member to the TimerBase class in order to do it. Please take a look.